### ISSUE TITLE
Catch-all errorhandler does not catch BadRequestKeyError

### ISSUE THREAD

**Issue by sampeka** [pallets/flask#2268]
## Versions

Python 3.6.1
Flask 0.12.1
Werkzeug 0.12.1

## Issue

A catch-all error handler (one that catches`Exception`), does not catch `werkzeug.exceptions.BadRequestKeyError`. Here's a very quick illustration of the issue:

https://github.com/sampeka/flask-issue

Steps to repoduce:
1. Clone the repo above, install requirements etc
2. Run `python ./example.py`
3. Navigate to `/good`, see that the error handler is correctly called 
4. Navigate to `/bad`, see that the error handler is not called 

If I define an error handler specifically for `BadRequestKeyError` it will be called, as expected.

Edit: If you downgrade to `Flask==0.10.1`, this example will work.

**Comment by sampeka** [comment]
Closing since this has the same issue: https://github.com/pallets/flask/issues/2267

**Comment by ThiefMaster** [comment]
Does enabling the `TRAP_BAD_REQUEST_ERRORS` option help

**Comment by sampeka** [comment]
@ThiefMaster no, still same issue

**Comment by sampeka** [comment]
Actually, turns out the issue is with other `werkzeug` exceptions as well. Not sure if it's the same issue as the one I linked above, so reopening.

**Comment by ngaya-ll** [comment]
This is more similar to #941.  This issue here is that `HTTPException`s are treated differently from other exceptions in the error handling code.  Each HTTP status code gets its own handler map.  So in order to catch `BadRequestKeyError` you would need to register the handler for the `404` code.

**Comment by Jane85** [comment]
Is there still a way now to catch all HTTPExceptions? I believe that this is a good feature as we might use third party libraries and it is not always clear what kind of exceptions are thrown by the third party library.

**Comment by ngaya-ll** [comment]
You can catch standard `HTTPException`s by iterating over werkzeug's standard error codes and registering a handler for each of them (can be the same handler):
```
from werkzeug.exceptions import default_exceptions
for code in default_exceptions:
        app.errorhandler(code)(handler)
```

This will work for most real-life cases.  However, there are several bugs/caveats:
* If an `HTTPException` is raised with a `code` attribute of `None`, it bypasses exception handling
* Error handlers registered with integer codes will only catch `HTTPException`s with that code if the exception class inherits from werkzeug's default exception class for that code.
* Custom `HTTPException`s with non-standard codes will not be caught.

You can change http exception handling by setting the `TRAP_HTTP_EXCEPTIONS` config parameter to `True`.  With this configuration:
* `HTTPException`s with a `code` of `None` will be handled like regular exceptions
* `HTTPException`s with other `code` values are handled the same as usual (i.e. per-code error handler lookup), except that if no handler is found the exception will be re-raised instead of generating a default error page.

**Comment by Jane85** [comment]
@ngaya-ll Hi, thanks for your quick reply! I tried the code that you suggested, however it does not take any effect. But I am using flask-restplus on top of flask, could it be that they have some conflicts?

**Comment by ngaya-ll** [comment]
I'm not familiar with flask-restplus but from the [documentation](http://flask-restplus.readthedocs.io/en/stable/errors.html) it appears that that extension modifies the error handling behavior, so my above comment may not apply.

**Comment by davidism** [comment]
Fixed by #2314.
